src/v/cluster/node_status_provider.cc

```diff
@@ -0,0 +1,235 @@
+                            });
+                      }).handle_exception([](const std::exception_ptr& e) {
+        vlog(
+          clusterlog.debug,
+          "Exception during node status provider tick: {}",
+          e);
+    });
+}
+
+ss::future<> node_status_provider::dispatch_node_status_requests() {
+    node_status_request request = {.sender_metadata = {.node_id = _self}};
+
+    std::vector<ss::future<>> futures;
+    futures.reserve(_discovered_peers.size());
+
+    for (const auto& peer_id : _discovered_peers) {
+        auto fut = do_node_status(peer_id, request);
+        futures.push_back(std::move(fut));
+    }
+
+    return ss::when_all_succeed(futures.begin(), futures.end());
+}
+
+ss::future<> node_status_provider::do_node_status(
+  model::node_id target, node_status_request r) {
+    auto gate_holder = _gate.hold();
+
+    auto reply = co_await send_node_status_request(target, r);
+    if (reply.has_error()) {
+        auto err = reply.error();
+        if (
+          err.category() == rpc::error_category()
+          && static_cast<rpc::errc>(err.value())
+               == rpc::errc::client_request_timeout) {
+            _stats.rpcs_timed_out += 1;
+        }
+
+        vlog(
+          clusterlog.debug,
+          "Error occurred while sending node status request: {}",
+          err.message());
+
+        co_return;
+    }
+
+    _stats.rpcs_sent += 1;
+    process_reply(std::move(reply));
+
+    co_return;
+}
+
+ss::future<result<node_status_reply>>
+node_status_provider::send_node_status_request(
+  model::node_id target, node_status_request r) {
+    auto broker = _members_table.get_broker(target);
+    if (!broker) {
+        co_return make_error_code(errc::node_does_not_exists);
+    }
+
+    if (!_node_connection_cache.contains(target)) {
+        auto cert = co_await maybe_build_reloadable_certificate_credentials(
+          _rpc_tls_config);
+        co_await _node_connection_cache.emplace(
+          target,
+          rpc::transport_configuration{
+            .server_addr = broker.value()->rpc_address(),
+            .credentials = cert,
+            .disable_metrics = net::metrics_disabled::yes},
+          rpc::make_exponential_backoff_policy<rpc::clock_type>(
+            std::chrono::seconds(1), std::chrono::seconds(5)));
+    }
+
+    auto send_by = rpc::clock_type::now() + _period();
+    co_return co_await _node_connection_cache
+      .with_node_client_locally<node_status_provider_rpc_client_protocol>(
+        target, send_by, [send_by, r = std::move(r)](auto client) mutable {
+            return client.node_status(std::move(r), rpc::client_opts(send_by))
+              .then(&rpc::get_ctx_data<node_status_reply>);
+        });
+}
+
+void node_status_provider::process_reply(result<node_status_reply> reply) {
+    vassert(ss::this_shard_id() == shard, "invoked on a wrong shard");
+
+    if (likely(reply)) {
+        update_node_status(std::move(reply.value().replier_metadata));
```

Comment:
I am wondering if we should use two timestapms. One storing last reply timestamp and the other one last request timestamp. This way we could leverage them to understand if other nodes have up to date information about its peers (responding to requests but not sending the requsts to current node)

Reply:
I thought about this more and I'm not too sure it would be useful. The status is updated when handling the request and when processing the reply, and since the request and response are symmetric I don't think the described scenario can happen. If a node is responding to requests it will have info about it's peers.

Reply:
what about the situation in which one of the nodes is in a state where it doesn't have information about a peer but the peer knows about that node (this can happen in case of corrupted raft0 log on one of the clusters or node being part of another cluster)

Reply:
If node A know about node B (because it sent a request and got a reply), then node B also knows about node A (it updated its status when it handled the request).